Have salt-minion deployed and configured on a new node

charlesprost opened the pull request:
Fixes #739 

The goal of this PR is to provide a way to have a new node connected to the salt-master through a freshly installed salt-minion.

Installing stuff on this new node to make it part of the control plane is not part of this PR.

Things you need to do to use this PR:

1. `vagrant destroy -f (or vagrant snapshot restore...) && ./doit.sh vagrantup`
1. Start a new node: `vagrant up node1`
1. (_could be done by default later_) Install `kubectl` on the bootstrap node and test it: `vagrant ssh bootstrap` then `sudo su` then `yum install -y kubectl --enablerepo="metalk8s-*" && export KUBECONFIG=/etc/kubernetes/admin.conf && kubectl get pods -n "kube-system"`
1. On the boostrap node, have a file describing node1: `vagrant ssh bootstrap` then `vi node1.yaml` then 
```
apiVersion: v1
kind: Node
metadata:
  name: node1
  labels:
    beta.kubernetes.io/os: linux
  annotations:
    metalk8s.scality.com/ssh-user: vagrant
    metalk8s.scality.com/ssh-port: "22"
    metalk8s.scality.com/ssh-host: <NODE1_IP_ADDRESS>
    metalk8s.scality.com/ssh-key-path: /etc/metalk8s/pki/preshared_key_for_k8s_nodes
    metalk8s.scality.com/ssh-sudo: "true"
```
while inserting the IP address of node1 and create it: `vagrant ssh bootstrap` then `sudo su` then `kubectl create -n "kube-system" -f node1.yaml`
1. On the boostrap node, connect to the salt-master container: `vagrant ssh bootstrap` then `sudo su` then `kubectl -n "kube-system" exec -it salt-master-bootstrap bash`
1. On the salt-master container, ping node1 to ensure SSH connection is OK: `salt-ssh -i node1 test.ping --roster=kubernetes`
1. On the salt-master container, deploy the salt-minion and accept its key: `salt-run state.orch metalk8s.orchestrate.deploy_salt_minion_on_new_node saltenv=metalk8s-2.0 pillar='{"bootstrap_id": "bootstrap", "node_name": "node1"}' ; salt-key -y -a node1`
1. Then wait a few sec and check it's working: `salt '*' test.ping`

Yet to do:

- [x] test if the roster needs to be synced before launch the orchestrate runner in the salt-master container
- [x] put the bootstrap ip (the one linked to the control plane network) into the mine during the bootstrap stage
- [x] use this ip when configuring the repos on the new node
- [x] use this ip when configuring the salt-minion on the new node from the salt-master container so the new salt-minion can reach the master
- [x] check how to pass the node name to the orchestrate target

NicolasT committed:
build, salt: bump Salt version and install new dependencies in `salt-master` image

Fixes #739

There is a bug in Salt 2018.3.3 related to using a roster from an orchestrate runner.

And we need salt-ssh, SSH clients, and a kubernetes client in order to run
our k8s roster and runner.

charlesprost committed:
vagrant: create a preshared SSH key when running make vagrantup

Fixes #739

In order to add a new Vagrant node to the cluster, we need to have
a SSH key pair.
We build one that has to be manually deployed into the nodes:
- the private part in the bootstrap node,
- the public part in the new node.

charlesprost committed:
salt: mount /etc/kubernetes into salt-master container

Fixes #739

In order to use the KubeAPI with a client, the salt-master container
needs an access to the KubeAPI configuration from the bootstrap node.

charlesprost committed:
salt: make metalk8s.repo generic enough to be used on a new node

Fixes #739

This state was too linked to the bootstrap node hence using it
on a new node was a failure because it tried to deploy containers
from the registry when we don't have anything alike in the new node.

charlesprost committed:
salt: add a mine state to register IPs on networks

Fixes #739

We want all minions to have access through the mine
to the bootstrap node IP on the control plane network.

bert-e commented:
# Hello nootal,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/install-salt-minion-on-a-new-node` into
`development/2.0`
* [w/2.1/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.1` 
* [w/2.2/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

gdemonet commented on Makefile:
```diff
@@ -278,8 +282,11 @@ $(ISO): $(ALL)
 
 vagrantup: $(ALL) ## Run 'vagrant up' to (re-)provision a development environment
+	mkdir -p .vagrant/
+	if [ ! -f $(VAGRANT_PRESHARED_KEY_PATH) ]; then ssh-keygen -t rsa -b 4096 -N '' -f $(VAGRANT_PRESHARED_KEY_PATH) -C Makefile ; fi
```
Maybe you could have this key defined as an independent target, e.g.:

```
vagrantup: $(ALL) $(VAGRANT_PRESHARED_KEY)
    $(VAGRANT) up $(VAGRANT_UP_OPTS)

$(VAGRANT_PRESHARED_KEY):
    mkdir -p $(@D)
    ssh-keygen -t rsa -b 4096 -N '' -f $@ -C Makefile
```

charlesprost replied:
Done. Thx for the tip.

gdemonet commented on salt/_roster/kubernetes_nodes_roster.py:
```diff
@@ -0,0 +1,52 @@
+    import kubernetes.config
+    HAS_DEPS = True
+except ImportError:
+    log.exception('Unable to load dependencies')
```
This log message might not be easily found, a good practice is to return a tuple `(False, "custom error message")` from `__virtual__` (see [here](https://docs.saltstack.com/en/latest/ref/modules/#returning-error-information-from-virtual))

charlesprost replied:
Done.

charlesprost committed:
salt: make metalk8s.salt.minion state generic enough to be used on a new node

Fixes #739

This state didn't allow to have a master located elsewhere that on localhost.

bert-e commented:
# History mismatch



Merge commit #7149cecc13c473fdd0c6d1a5bf62e996659f79c4 on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

charlesprost committed:
salt: add a kubernetes roster for nodes

Fixes #739

In order to get information from the KubeAPI, we needed a roster
able to talk to it.

charlesprost committed:
salt: add an orchestrate runner able to install salt-minion on a new node

Fixes #739

To install a salt-minion on a new node, we needed an orchestrate runner
that will be run from the salt master.

Execute on the salt-master container:

$ salt-run state.orch metalk8s.orchestrate.deploy_salt_minion_on_new_node saltenv=metalk8s-2.0 pillar='{"bootstrap_id": "bootstrap", "node_name": "node1"}' && salt-key -y -a node1

bert-e commented:
# History mismatch



Merge commit #d0dffbd9c31f3d7d546678f64c1689c0769e14ac on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

charlesprost commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/install-salt-minion-on-a-new-node` into
`development/2.0`
* [w/2.1/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.1` 
* [w/2.2/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

bert-e commented:
# History mismatch



Merge commit #23182494a4bb7a0b6a124fc8a2d1a893b8ffb97e on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_salt_minion_on_new_node.sls:
```diff
@@ -0,0 +1,23 @@
+{%- set control_plane_ips = salt.saltutil.runner('mine.get', tgt='*', fun='control_plane_ips') %}
+{%- if 'bootstrap' in control_plane_ips.keys() and control_plane_ips['bootstrap'] %}
+{%- set control_plane_ip = control_plane_ips['bootstrap'][0] %}
```
Wow "bootstrap" really hardcoded everywhere

NicolasT replied:
We can't even assume this is the correct host/minion name...

charlesprost replied:
I'll fix it as Christian has been doing it.

TeddyAndrieux commented on salt/metalk8s/mine/ips.sls:
```diff
@@ -0,0 +1,8 @@
+{%- from "metalk8s/map.jinja" import networks with context %}
+
+Advertise control plane network IPs in the mine:
```
Suggest to use a mine_function directly in the pillar instead of a state that push in the mine.

Can you check with @sayf-eddine-scality he asked me yesterday (and dit it btw) to have repos ip in the mine.

Quick explaination have a custom module that return the current ip (or hostname) ONLY IF the host is a control plane so maybe check pods or directly ask api server. 
And use this custom module as a mine_function so when we do `mine.update` we advertise all REAL control plane nodes automatically

PS: It's really simple to do

NicolasT replied:
+1 Wasn't aware that's possible :blush:

charlesprost replied:
Yeah that was something I tried to do but didn't success. I'll create a debt ticket.

charlesprost replied:
#833

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_salt_minion_on_new_node.sls:
```diff
@@ -0,0 +1,23 @@
+{%- set control_plane_ip = 'localhost' %}
+{%- endif %}
+
+Test:
```
`Test` ?

charlesprost replied:
Yeaaaaaaaaaaaaaaaaah OK

TeddyAndrieux reviewed (dismissed):
Not tested, nothing look critical feel free to dismiss if you want to do it later on

NicolasT commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+    client = None
+    try:
+        client = kubernetes.config.new_client_from_config(
+            config_file='/etc/kubernetes/admin.conf',
```
Please file a debt ticket to change the `kubeconfig` file being used.

charlesprost replied:
#834

NicolasT commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+            'host': item.metadata.annotations.get('metalk8s.scality.com/ssh-host', item.metadata.name),
+            'port': int(item.metadata.annotations.get('metalk8s.scality.com/ssh-port', 22)),
+            'user': item.metadata.annotations.get('metalk8s.scality.com/ssh-user', 'root'),
+            'priv': item.metadata.annotations.get('metalk8s.scality.com/ssh-priv-key-path', 'salt-ssh.rsa'),
```
Consider `private` instead of `priv`, or even drop `priv` (should be rather clear what it should be, and documented).

charlesprost replied:
Not sure what you meant, I can't find this in the doc :https://docs.saltstack.com/en/latest/topics/ssh/roster.html

NicolasT replied:
I was referring to the `priv` in the annotation key, not the field in the roster dict :)

charlesprost replied:
Oh OK :-P

NicolasT commented on salt/metalk8s/orchestrate/deploy_salt_minion_on_new_node.sls:
```diff
@@ -0,0 +1,23 @@
+  salt.state:
+    - ssh: true
+    - roster: kubernetes
+    - tgt: node1
```
Should be passed in as an argument to `salt-run` (through pillar values).

charlesprost replied:
Done.

NicolasT commented on salt/metalk8s/orchestrate/deploy_salt_minion_on_new_node.sls:
```diff
@@ -0,0 +1,23 @@
+    - ssh: true
+    - roster: kubernetes
+    - tgt: node1
+    - saltenv: metalk8s-2.0
```
Same as https://github.com/scality/metalk8s/pull/815/files#r268107408

charlesprost replied:
Done.

NicolasT reviewed (dismissed):
Some comments. Most of them may, however, be tackled in follow-up PRs. In that case, please either add tasks to the existing ticket, or file new tickets (either debt or not, in the current milestone then).

bert-e commented:
# History mismatch



Merge commit #68ed52a52a257d76b57d5805939dcb9855043cac on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

charlesprost commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #68ed52a52a257d76b57d5805939dcb9855043cac on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

charlesprost commented:
PR is ready. At the end we have a working minion on the new node.

I'll work on doit integration for the part in the boostrap and then we can decide if we merge it before or after the PR from @slaperche-scality

charlesprost commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/install-salt-minion-on-a-new-node` with contents from `improvement/install-salt-minion-on-a-new-node`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/install-salt-minion-on-a-new-node origin/development/2.1
 $ git merge origin/improvement/install-salt-minion-on-a-new-node
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/install-salt-minion-on-a-new-node
```





*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/install-salt-minion-on-a-new-node` into
`development/2.0`
* [w/2.1/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.1` 
* [w/2.2/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux







*The following options are set:* **approve**

charlesprost commented:
> # Conflict
> 
> A conflict has been raised during the creation of
> integration branch `w/2.1/improvement/install-salt-minion-on-a-new-node` with contents from `improvement/install-salt-minion-on-a-new-node`
> and `development/2.1`.
> 
> **I have not created the integration branch.**
> 
> Here are the steps to resolve this conflict:
> 
> ```shell
>  $ git fetch
>  $ git checkout -B w/2.1/improvement/install-salt-minion-on-a-new-node origin/development/2.1
>  $ git merge origin/improvement/install-salt-minion-on-a-new-node
>  $ # <intense conflict resolution>
>  $ git commit
>  $ git push -u origin w/2.1/improvement/install-salt-minion-on-a-new-node
> ```
> 
> _The following options are set:_ **approve**

Done.

slaperche-scality commented:
> then we can decide if we merge it before or after the PR from @slaperche-scality

Mine is on its way to be merged and should land this evening so I guess you can merge after mine.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/install-salt-minion-on-a-new-node` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/install-salt-minion-on-a-new-node`).

```bash
 $ git fetch
 $ git checkout origin/improvement/install-salt-minion-on-a-new-node
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/install-salt-minion-on-a-new-node
```







*The following options are set:* **approve**

slaperche-scality commented on packages/yum_repositories/saltstack.repo:
```diff
@@ -1,6 +1,6 @@
 [saltstack]
 name=SaltStack repo for RHEL/CentOS $releasever
-baseurl=https://repo.saltstack.com/yum/redhat/$releasever/$basearch/archive/2018.3.3
+baseurl=https://repo.saltstack.com/yum/redhat/$releasever/$basearch/archive/2018.3.4
```
Out of the scope of this PR, but shouldn't we "templatize" all those use of the Salt version (using `SALT_MASTER_IMAGE_SALT_VERSION` which is defined in our buildchain)?

Ideally, bumping the Salt version would require to edit only one of two place and not many files like we currently have to do (it's error prone).

NicolasT replied:
#640 

I think the 'best' approach, especially given with `doit` we can use whichever input (format) we want (which is less the case with `make`) is to have a YAML file in the `salt` tree which can be imported in SLS files to retrieve all versions, and can also be read by the `doit` modules to do whatever they want. We have more flexibility in the `doit` modules than in the Salt files, hence embedding in Salt and consuming from `doit` rather than embedding in `doit` and somehow embedding in Salt, unless we'd create a Salt-readable file from the `doit` data in some step, but that seems like a useless indirection to me.

slaperche-scality replied:
> #640

Ha we already have an issue for that, perfect!

> I think the 'best' approach, especially given with `doit` we can use whichever input (format) we want (which is less the case with `make`) is to have a YAML file in the `salt` tree which can be imported in SLS files to retrieve all versions, and can also be read by the `doit` modules to do whatever they want.

We would need to add a YAML parser to the build virtualenv, but that's not a big dependency (and it's better than generating Salt-readable file, as you mentioned, which would probably require a YAML lib as well anyway).

charlesprost replied:
You may want to discuss this in the #640 instead ;-)

slaperche-scality commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+import fnmatch
+import logging
+log = logging.getLogger(__name__)
```
Skip a line between the import and the constants declaration.

In fact I would even move the log declaration lower since here you're in the middle of the import statements.

charlesprost replied:
Done.

slaperche-scality commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+    if HAS_DEPS:
+        return HAS_DEPS
+    else:
+        return False, 'Unable to load the kubernetes nodes roster: dependencies unavailable.'
```
minor: could add the names in the error message for an easier troubleshooting.

charlesprost replied:
Done.

slaperche-scality commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+
+    v1 = kubernetes.client.CoreV1Api(api_client=client)
+
+    nodes = None
```
The variables can be declared l.39, no?

And if you want a default value, then `[]` is better than `None` since you're iterating over it below.

charlesprost replied:
Just removed the default declaration since we don't need them.

slaperche-scality commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+            'host': item.metadata.annotations.get('metalk8s.scality.com/ssh-host', item.metadata.name),
+            'port': int(item.metadata.annotations.get('metalk8s.scality.com/ssh-port', 22)),
+            'user': item.metadata.annotations.get('metalk8s.scality.com/ssh-user', 'root'),
+            'priv': item.metadata.annotations.get('metalk8s.scality.com/ssh-priv-key-path', 'salt-ssh.rsa'),
```
Lines are way too long…

You could declare an intermediate variable `annotations = item.metadata.annotations` and use that instead (yeah, you may have to drop to a `loop` instead of a dict comprehension but the readability gain is worth it IMHO).

charlesprost replied:
Done.

bert-e commented:
# History mismatch



Merge commit #b5c3495a2c43b97307f361492ad9753de7974b3e on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

NicolasT commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,55 @@
+            'port': int(item.metadata.annotations.get('metalk8s.scality.com/ssh-port', 22)),
+            'user': item.metadata.annotations.get('metalk8s.scality.com/ssh-user', 'root'),
+            'priv': item.metadata.annotations.get('metalk8s.scality.com/ssh-priv-key-path', 'salt-ssh.rsa'),
+            'sudo': True
```
Unless username is `root`, which we have as the fallback default? Or, also make this configurable as an annotation?

charlesprost replied:
Yep let's move it to annotations ;-)

NicolasT reviewed (dismissed):
Would be nice to, over time, deploy the SSH keypair on the Vagrant hosts by default (a `shell` provisioning step in the `Vagrantfile`).

NicolasT commented:
We'll also want to test this in CI at some point in time, of course ;-)

charlesprost commented:
@NicolasT yes to your two comments, it wil be part of the testing around this control plane expansion feature

charlesprost commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/install-salt-minion-on-a-new-node` with contents from `improvement/install-salt-minion-on-a-new-node`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/install-salt-minion-on-a-new-node origin/development/2.1
 $ git merge origin/improvement/install-salt-minion-on-a-new-node
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/install-salt-minion-on-a-new-node
```





*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/install-salt-minion-on-a-new-node` into
`development/2.0`
* [w/2.1/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.1` 
* [w/2.2/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/16ec3bfa34c27b7ead3541d6c92dffe8a5813108) did not succeed in branch improvement/install-salt-minion-on-a-new-node.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #16ec3bfa34c27b7ead3541d6c92dffe8a5813108 on the integration branch
`w/2.1/improvement/install-salt-minion-on-a-new-node` is merging a branch which is neither the current
branch `improvement/install-salt-minion-on-a-new-node` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/install-salt-minion-on-a-new-node` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

charlesprost commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/install-salt-minion-on-a-new-node` with contents from `improvement/install-salt-minion-on-a-new-node`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/install-salt-minion-on-a-new-node origin/development/2.1
 $ git merge origin/improvement/install-salt-minion-on-a-new-node
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/install-salt-minion-on-a-new-node
```





*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/install-salt-minion-on-a-new-node` into
`development/2.0`
* [w/2.1/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.1` 
* [w/2.2/improvement/install-salt-minion-on-a-new-node](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/install-salt-minion-on-a-new-node) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/66d11b4202a07ef301d86437b2703aa84297d2f8) did not succeed in branch w/2.2/improvement/install-salt-minion-on-a-new-node.




*The following options are set:* **approve**

slaperche-scality commented on buildchain/buildchain/constants.py:
```diff
@@ -28,6 +28,13 @@
 REPO_ROOT : Path = ISO_ROOT/'packages'
 # Root for the packages that we build ourselves.
 PKG_ROOT : Path = config.BUILD_ROOT/'packages'
+# Root of the Vagrant environment folder
```
Minor: period at the end of the sentence

slaperche-scality commented on salt/_roster/kubernetes_nodes.py:
```diff
@@ -0,0 +1,61 @@
+        return HAS_DEPS
+    else:
+        return False, 'Unable to load the kubernetes nodes roster: '\
+            'dependencies kubernetes.client and/or '\
```
minor: "Kubernetes Python client library" may have been clearer than "kubernetes.client and/or kubernetes.config"

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nootal.
